Assignment on return sometimes generates bad C code

While porting a XORshift PRNG to ooc, I ran into an issue where the following code generated erroneous C code, causing gcc to fail:

``` ooc
get32 : func -> UInt32
{
   tmp := state32 [0] ^ (state32 [0] << 11)
   state32 [0] = state32 [1]
   state32 [1] = state32 [2]
   state32 [2] = state32 [3]

   return state32 [3] = (state32 [3] ^ (state32 [3] >> 19)) ^ (tmp ^ (tmp >> 8))
}
```

GCC prints out:

```
rock_tmp/./xorshift.c: In function ‘xorshift__xorgen_init_withSeed’:
rock_tmp/./xorshift.c:29:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c:30:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c:31:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c:32:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c:33:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c:34:5: warning: suggest parentheses around ‘+’ inside ‘<<’ [-Wparentheses]
rock_tmp/./xorshift.c: In function ‘xorshift__xorgen_get32_impl’:
rock_tmp/./xorshift.c:46:37: error: expected expression before ‘if’
rock_tmp/./xorshift.c:42:26: warning: unused variable ‘tmp’ [-Wunused-variable]
rock_tmp/./xorshift.c:47:1: warning: control reaches end of non-void function [-Wreturn-type]
```

--- a/sdk/lang/Array.h
+++ b/sdk/lang/Array.h
@@ -1,4 +1,3 @@
-
 #ifndef ___lang_array___
 #define ___lang_array___
 
@@ -22,11 +21,10 @@
     ((type* restrict) array.data)[index])
 
 #define _lang_array__Array_set(array, index, type, value) \
-    if(index < 0 || index >= array.length) { \
-        lang_Exception__Exception_throw((lang_Exception__Exception *) lang_Exception__OutOfBoundsException_new_noOrigin(index, array.length)); \
-        exit(1); \
-    } \
-    ((type* restrict) array.data)[index] = value;
+    (index < 0 || index >= array.length) ? \
+    lang_Exception__Exception_throw((lang_Exception__Exception *) lang_Exception__OutOfBoundsException_new_noOrigin(index, array.length)), \
+    *((type*) NULL) : \
+    (((type* restrict) array.data)[index] = value)
 
 #define _lang_array__Array_free(array) { array_free(array.data) }
 

